### PR TITLE
I'd like to be able to publish while running builds in parallel.

### PR DESCRIPTION
**Pull request by splatteredbits**
Created a task that does its best to wait for other jobs in a build to complete before continuing to the next build task. There are limitations because AppVeyor's API only returns information for the most recently started builds, but this is better than nothing.

### CODE REVIEW

File: appveyor.yml [hunk 1]
```diff
@@ -6,11 +6,6 @@ matrix:
   fast_finish: true
 
 environment:
-  NVM_HOME: C:\Users\appveyor\AppData\Roaming\nvm
-  POWERSHELL_GALLERY_API_KEY:
-    secure: Y7B8n8Ott858qlg1xt1Kz7VnvR59aO963p5dquIEQsMUZpvK/Ip5Rmj7G/jBKM6r
-  GITHUB_ACCESS_TOKEN:
-    secure: cEpuAn4TbJ17Y4WMMsHj46T4DLFMGE4n7xafkbiVQL4jueqc9TGG7NHWg4qbwmt7
```
Comment (DecoyJoe) [inline appveyor.yml]: Why were these removed? Are they now being set outside of source control, in the AppVeyor console?

File: Whiskey/Tasks/AppVeyorWaitForBuildJobs.ps1 [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+        [TimeSpan]$ReportInterval = '00:01:00',
+
+        [string[]]$InProgressStatus = @('running','queued'),
+
+        [string[]]$SuccessStatus = @('success')
+    )
+
+    Set-StrictMode -Version 'Latest'
+    Use-CallerPreference -Cmdlet $PSCmdlet -SessionState $ExecutionContext.SessionState
```
Comment (DecoyJoe) [inline Whiskey/Tasks/AppVeyorWaitForBuildJobs.ps1]: Maybe add an `$ErrorActionPreference = 'Stop'` to this task, so that if anything unexpected happens we fail the build rather than the task short-circuiting continuing with the rest of the publishing tasks.

File: Test/AppVeyorWaitForBuildJobs.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+'@
+        GivenJob -WithStatus 'queued' -ThatFinishesAtCheck 2 -WithFinalStatus 'failed' -ThatHasFinishedProperty
+        GivenJob -Current 
+        GivenLatestBuildIDIsNotThisBuild
+        WhenRunningTask -ErrorAction SilentlyContinue
+        ThenFails
+        $Global:Error | Should -Match 'unable to wait'
+        ThenCheckedStatus -Times 1 -Exactly
+    }
+}
```
Comment (DecoyJoe) [inline Test/AppVeyorWaitForBuildJobs.Tests.ps1]: Add a test when the `ApiKeyID` property is not specified.

File: Whiskey/Tasks/AppVeyorWaitForBuildJobs.ps1 [hunk 4]
```diff
@@ -0,0 +1,99 @@
+
+function Wait-WhiskeyAppVeyorBuildJob
+{
+    [CmdletBinding()]
+    [Whiskey.Task('AppVeyorWaitForBuildJobs')]
+    param(
+        [Parameter(Mandatory)]
+        [Whiskey.Context]$TaskContext,
+
+        [Parameter(Mandatory)]
```
Comment (DecoyJoe) [inline Whiskey/Tasks/AppVeyorWaitForBuildJobs.ps1]: Setting a task property as mandatory via this attribute currently won't work without additional changes to `Invoke-WhiskeyTask` or `Get-TaskParameter` I think. If someone does not specify this attribute in their `whiskey.yml` then the build will hang as PowerShell prompts the user for this mandatory parameter value.

File: Whiskey/Tasks/AppVeyorWaitForBuildJobs.ps1 [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+        $failedJobs = $jobsToCheck | Where-Object { $_.status -notin $SuccessStatus }
+
+        if( $failedJobs )
+        {
+            $suffix = 's'
+            if( ($failedJobs | Measure-Object).Count -gt 1 )
+            {
+                $suffix = 's'
+            }
```
Comment (DecoyJoe) [inline Whiskey/Tasks/AppVeyorWaitForBuildJobs.ps1]: Typo, first `$suffix` should be `$suffix = ''`

File: Test/AppVeyorWaitForBuildJobs.Tests.ps1 [hunk 6]
```diff
@@ -0,0 +1,474 @@
...
+        {
+            $script:currentJobID = $jobID
+        }
+
+        $job = [pscustomobject]@{ 
+            'name' = ('job{0}' -f $jobID)
+            'jobId' = $jobID;
+            'status' = $WithStatus;
+            'checks' = $ThatFinishesAtCheck;
+            'finalStatus' = $WithFinalStatus;
```
Comment (DecoyJoe) [inline Test/AppVeyorWaitForBuildJobs.Tests.ps1]: Nit: Unnecessary semicolons.

File: Test/AppVeyorWaitForBuildJobs.Tests.ps1 [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+        $Value
+    )
+
+    $script:whiskeyYml = $Value 
+}
+
+function Init
+{
+    $script:failed = $false
+    $script:testRoot = $TestDrive.FullName
```
Comment (DecoyJoe) [inline Test/AppVeyorWaitForBuildJobs.Tests.ps1]: Use `New-WhiskeyTestRoot` ?

File: Test/AppVeyorWaitForBuildJobs.Tests.ps1 [hunk 8]
```diff
@@ -0,0 +1,474 @@
...
+    $script:secret = $null
+    $script:secretID = $null
+    $script:whiskeyYml = $null
+    $script:jobs = $null
+    $script:currentJobID = $null
+    $script:runByAppVeyor = $null
+    $script:latestBuildID = $null
+    Remove-Item -Path (JOin-Path -Path $testRoot -ChildPath 'whiskey.yml') -ErrorAction Ignore
```
Comment (DecoyJoe) [inline Test/AppVeyorWaitForBuildJobs.Tests.ps1]: Typo: `JOin-Path`

File: Test/AppVeyorWaitForBuildJobs.Tests.ps1 [hunk 9]
```diff
@@ -0,0 +1,474 @@
...
+    ApiKeyID: AppVeyor
+    CheckInterval: 00:00:00.1
+    ReportInterval: 00:00:00.1
+'@
+        GivenJob -WithStatus 'running' -ThatFinishesAtCheck 2 -WithFinalStatus 'success'
+        GivenJob -Current 
+        GivenJob -WithStatus 'running' -ThatFinishesAtCheck 10 -WithFinalStatus 'success'
+        WhenRunningTask
+        ThenSucceeds
+        ThenCheckedStatus -Times 10
```
Comment (DecoyJoe) [inline Test/AppVeyorWaitForBuildJobs.Tests.ps1]: Also add a test that ensures the task is sleeping for the specified interval? It looks like this just ensures the task iterates a number of times while checking jobs.

File: appveyor.yml (hunk 1 above)
Comment (splatteredbits) [reply appveyor.yml]: AppVeyor merges what's in the UI with what's in your `appveyor.yml` so these aren't actually needed here.

File: Whiskey/Tasks/AppVeyorWaitForBuildJobs.ps1 (hunk 2 above)
Comment (splatteredbits) [reply Whiskey/Tasks/AppVeyorWaitForBuildJobs.ps1]: 🤔

File: Test/AppVeyorWaitForBuildJobs.Tests.ps1 (hunk 6 above)
Comment (splatteredbits) [reply Test/AppVeyorWaitForBuildJobs.Tests.ps1]: Ack! I like my semicolons in hashtables. From my cold dead hands!

File: Whiskey/Tasks/AppVeyorWaitForBuildJobs.ps1 (hunk 2 above)
Comment (splatteredbits) [reply Whiskey/Tasks/AppVeyorWaitForBuildJobs.ps1]: Any task could be used in the Publish portion of a pipeline. Shouldn't we have `$ErrorActionPreference = 'Stop'` in every task for the same reason?

Comment (DecoyJoe) [reply Whiskey/Tasks/AppVeyorWaitForBuildJobs.ps1]: Yes, I agree we should have that for all tasks and I actually thought we already did in `Invoke-WhiskeyTask` but turns out we don't. We mostly rely on calling `Stop-WhiskeyTask` to throw an exception during the build. I wasn't sure if there was a historical reason or not why we aren't using `$ErrorActionPreference = 'Stop'` for all tasks.

Comment (splatteredbits) [reply Whiskey/Tasks/AppVeyorWaitForBuildJobs.ps1]: Because it's PowerShell and not all errors are errors worth stopping for.

**Comment by DecoyJoe** [comment]
@splatteredbits The Whiskey [Wiki](https://github.com/webmd-health-services/Whiskey/wiki) still needs updating with documentation for the new `AppVeyorWaitForBuildJobs` task.
